aep/general/0004/aep.md

```diff
@@ -0,0 +1,101 @@
+becomes important to have a single, globally-unique type name. Additionally,
+tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
+
+## Guidance
+
+APIs **must** define a resource type for each resource in the API, according to
+the following pattern: `{API Name}/{Type}`. The type name:
+
+- **must** Start with an uppercase letter.
+- **must** Only contain alphanumeric characters.
+- **must** Be of the singular form of the noun.
+- **must** Use PascalCase (UpperCamelCase).
+- For Kubernetes, the type name **must** match the [object][] name.
+- For Protobuf, the type name **must** match the name of the protobuf message.
+
+### Examples
+
+Examples of resource types include:
+
+- `pubsub.googleapis.com/Topic`
+- `pubsub.googleapis.com/Subscription`
+- `spanner.googleapis.com/Database`
+- `spanner.googleapis.com/Instance`
+- `networking.istio.io/Instance`
+
+### Annotating resource types
+
+APIs **must** annotate the resource types for each resource in the API using
+the [`google.api.resource`][resource] annotation:
+
+```proto
+// A representation of a Pub/Sub topic.
+message Topic {
+  option (google.api.resource) = {
+    type: "pubsub.googleapis.com/Topic"
+    pattern: "projects/{project}/topics/{topic}"
+    singular: "topic"
+    plural: "topics"
+  };
+
+  // Name and other fields...
+}
+```
+
+- Patterns **must** correspond to the [resource path][resource-paths].
+- Pattern variables (the segments within braces) **must** use `snake_case`, and
+  **must not** use an `_id` suffix.
+- Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.
+- Pattern variables **must** be unique within any given pattern. (e.g.
+  `projects/{abc}/topics/{abc}` is invalid; this is usually a natural corollary
+  of collection identifiers being unique within a pattern.)
+- Singular **must** be the lower camel case of the type.
+  - Pattern variables **must** be the singular form of the resource type e.g. a
+    pattern variable representing a `Topic` resource ID is named `{topic}`.
+- Plural **must** be the lower camel case plural of the singular.
+  - Pattern collection identifier segments **must** match the plural of the
+    resources, except in the case of [nested collections][].
+
+#### Pattern uniqueness
+
+When multiple patterns are defined within a resource, these patterns **must**
+be mutually unique, where uniqueness is defined as being by-character identical
+once all resource ID path segments have been removed, leaving all `/`
+separators.
+
+Therefore the following two patterns **must not** be defined within the same
+resource:
+
+- `user/{user}`
+- `user/{user_part_1}~{user_part_2}`
+
+## Rationale
+
+### Singular and Plural
+
+Well-defined singular and plurals of a resource enable clients to determine the
+proper name to use in code and documentation.
+
+lowerCamelCase can be translated into other common forms of a resource path
+such as UpperCamelCase and snake_case.
```

Comment:
I don't know exactly what this is the rationale for...and all the string cases can be converted to the others, right?

Reply:
I think the main issue was acronyms - e.g. if you have a resource name APIMethod for example - how would one know that should be translated to API_method?

Although I guess should we just have the type be in kebab-case, to match the resource path?

Reply:
updated to kebab-case.